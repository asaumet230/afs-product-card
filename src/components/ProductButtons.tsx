import React from 'react';
import { CSSProperties, useCallback, useContext } from 'react';

// Context:
import { ProductContext } from "../context/productContext";

// CSS:
import styles from '../styles/styles.module.css';



export interface Props {
    className?: string;
    style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {


    const { counter, increaseBy, maxCount } = useContext(ProductContext);

    const isMaxReached = useCallback(() => (!!maxCount && maxCount === counter), [maxCount, counter]);

    return (
        <div
            style={style}
            className={`${styles.buttonsContainer} ${className}`}>
            <button
                className={styles.buttonMinus}
                onClick={() => increaseBy(-1)}>
                -
            </button>

            <div className={styles.countLabel}>
                {counter}
            </div>

            <button
                className={`${isMaxReached() && styles.disabled} ${styles.buttonAdd}`}
                onClick={() => increaseBy(1)}
                disabled={isMaxReached()}
            >
                +
            </button>
        </div>
    )
}