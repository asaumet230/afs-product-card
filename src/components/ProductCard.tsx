import React from 'react';
import { CSSProperties } from 'react';

// Context:
import { ProductContext } from '../context/productContext';

// Hooks:
import { useProduct } from '../hooks/useProduct';

// CSS:
import styles from '../styles/styles.module.css';

// Interfaces:
import { InitialValues, OnChangeArgs, Product, ProductCartHandlers } from '../interfaces/interfaces';



export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[];
    children: (args: ProductCartHandlers) => JSX.Element
    className?: string;
    style?: CSSProperties;
    onChange?: (args: OnChangeArgs) => void;
    value?: number;
    initialValues?: InitialValues;
}



export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

    const { Provider } = ProductContext;
    const { counter, increaseBy, reset, maxCount, isMaxCountReached } = useProduct({
        onChange,
        product,
        value,
        initialValues
    });


    return (

        <Provider value={{
            counter,
            product,
            maxCount,
            increaseBy,
        }}>
            <div
                style={style}
                className={`${styles.productCard} ${className}`}>
                {children({
                    product,
                    increaseBy,
                    reset,
                    maxCount,
                    count: counter,
                    isMaxCountReached,
                })}
            </div>
        </Provider>


    )
}

