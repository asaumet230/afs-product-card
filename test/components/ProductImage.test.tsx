import React from "react";
import { create } from "react-test-renderer";

// Component:
import { ProductImage, ProductCard } from '../../src/components';


// Product:
import { product1, product2 } from '../data/products';

describe('Pruebas en el componente <ProductImage />', () => {

    test('Debe de renderizarse correctamente sin imagen', () => {

        const wrapper = create(
            <ProductCard product={product1}>
                {
                    () => (
                        <ProductImage img="http://no-image.jpg" />
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();


    });


    test('Debe de renderizarse correctamente con imagen', () => {

        const wrapper = create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

    });

});