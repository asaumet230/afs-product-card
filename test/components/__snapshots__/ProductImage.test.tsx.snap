// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <ProductImage /> Debe de renderizarse correctamente con imagen 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffe-con-imagen"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`Pruebas en el componente <ProductImage /> Debe de renderizarse correctamente sin imagen 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffe-sin-imagen"
    className="productImg undefined"
    src="http://no-image.jpg"
  />
</div>
`;
